core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
If an impl is doing that and it isn't thread-safe, surely that's a busted implementation? you can't expect that no two instances of the class in the same JVM would not be accessed concurrently (unless you design for it of course). At the least, no reading of data happens in this path anyway? I'm trying to avoid designing around this, since this is a flag that would be on by default, hidden, but it's not clear how someone would reasonably figure out this was the problem and this was the solution.